source/snapshot/iterator.go

```diff
@@ -0,0 +1,173 @@
+	Snapshot   string
+}
+
+type Iterator struct {
+	db      *pgxpool.Pool
+	t       *tomb.Tomb
+	workers []*FetcherWorker
+
+	conf Config
+
+	lastPosition position.Position
+
+	records chan sdk.Record
+}
+
+func NewIterator(ctx context.Context, db *pgxpool.Pool, c Config) (*Iterator, error) {
+	p, err := position.ParseSDKPosition(c.Position)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse position: %w", err)
+	}
+
+	if p.Snapshot == nil {
+		p.Snapshot = make(position.SnapshotPositions)
+	}
+
+	i := &Iterator{
+		db:           db,
+		t:            &tomb.Tomb{},
+		conf:         c,
+		records:      make(chan sdk.Record),
+		lastPosition: p,
+	}
+
+	if err := i.initFetchers(ctx); err != nil {
+		return nil, fmt.Errorf("failed to initialize table fetchers: %w", err)
+	}
+
+	go i.start(ctx)
+
+	return i, nil
+}
+
+func (i *Iterator) Next(ctx context.Context) (sdk.Record, error) {
+	select {
+	case <-ctx.Done():
+		return sdk.Record{}, fmt.Errorf("iterator stopped: %w", ctx.Err())
+	case r, ok := <-i.records:
+		if !ok { // closed
+			if err := i.t.Err(); err != nil {
+				return sdk.Record{}, fmt.Errorf("fetchers exited unexpectedly: %w", err)
+			}
+			return sdk.Record{}, ErrIteratorDone
+		}
+
+		if err := i.updateLastPosition(&r); err != nil {
+			return sdk.Record{}, fmt.Errorf("failed to update last fetched position: %w", err)
+		}
+
+		return r, nil
+	}
+}
+
+func (i *Iterator) Ack(_ context.Context) error {
+	return nil
+}
+
+func (i *Iterator) Teardown(_ context.Context) error {
+	if i.t != nil {
+		i.t.Kill(errors.New("tearing down snapshot iterator"))
+	}
+
+	return nil
+}
+
+func (i *Iterator) updateLastPosition(r *sdk.Record) error {
+	pos, err := position.ParseSDKPosition(r.Position)
+	if err != nil {
+		return fmt.Errorf("failed to parse position: %w", err)
+	}
+
+	// merge this position with latest position
+	i.lastPosition.Type = pos.Type
+	for k, v := range pos.Snapshot {
+		i.lastPosition.Snapshot[k] = v
+	}
+
+	r.Position = i.lastPosition.ToSDKPosition()
+
+	return nil
+}
+
+func (i *Iterator) initFetchers(ctx context.Context) error {
+	var errs []error
+
+	i.workers = make([]*FetcherWorker, len(i.conf.Tables))
+
+	for j, t := range i.conf.Tables {
+		w := NewFetcherWorker(i.db, i.records, FetcherConfig{
+			Table:    t,
+			Key:      i.conf.TablesKeys[t],
+			Snapshot: i.conf.Snapshot,
+			Position: i.lastPosition,
+		})
+
+		if err := w.Validate(ctx); err != nil {
+			errs = append(errs, fmt.Errorf("failed to validate table fetcher %q config: %w", t, err))
+		}
+
+		i.workers[j] = w
+	}
+
+	if len(errs) != 0 {
+		return errors.Join(errs...)
+	}
+
+	return nil
+}
+
+func (i *Iterator) start(ctx context.Context) {
+	var wg sync.WaitGroup
+	for j := range i.workers {
+		f := i.workers[j]
+
+		wg.Add(1)
+		i.t.Go(func() error {
+			ctx := i.t.Context(ctx)
+			defer wg.Done()
+
+			if err := f.Run(ctx); err != nil {
+				return fmt.Errorf("fetcher exited: %w", err)
+			}
+
+			return nil
+		})
+	}
+	defer close(i.records)
```

Comment:
When fetchers terminate, we might still have some records in this channel. WDYT about draining the channel before actually calling it done?

Reply:
Does that need to happen?  My thinking is this:

1. Channel is not buffered, so fetchers will wait on injecting records as long as next is called.
2. If fetchers are done, then there is no more data coming through the channel, which means they are not blocking on sending data
3. Once all fetchers have exited the channel can be closed, since should be empty.

Now if the channel is buffered, Next() should be able to drain the channel  it hits the not ok when receiving. 
Even if there are any messages in the channel and they get dropped because of teardown or cancelation, the position will not be updated which means we can safely resume from when the last position was saved.

Does that make sense?

Reply:
Yup, makes sense, I missed that the channel was unbuffered, that helped a lot.:)

Reply:
Sorry @hariso, didn't reload the page when I posted this comment. GitHub should really implement live loading of new comments, this happened a few times already 😞